runtime/near-vm-runner/src/runner.rs

```diff
@@ -48,59 +24,57 @@ pub fn run<'a>(
-        );
-    }
-    if method_name.is_empty() {
-        return (
-            None,
-            Some(VMError::FunctionCallError(FunctionCallError::MethodResolveError(
-                MethodResolveError::MethodEmptyName,
-            ))),
-        );
-    }
-
-    let module = match cache::compile_module(code_hash, code, wasm_config) {
-        Ok(x) => x,
-        Err(err) => return (None, Some(err)),
-    };
-    let mut memory = match WasmerMemory::new(
-        wasm_config.limit_config.initial_memory_pages,
-        wasm_config.limit_config.max_memory_pages,
-    ) {
-        Ok(x) => x,
-        Err(_err) => panic!("Cannot create memory for a contract call"),
-    };
-    let memory_copy = memory.clone();
-
-    let mut logic =
-        VMLogic::new(ext, context, wasm_config, fees_config, promise_results, &mut memory);
-
-    let import_object = imports::build(memory_copy, &mut logic);
-
-    let method_name = match std::str::from_utf8(method_name) {
-        Ok(x) => x,
-        Err(_) => {
-            return (
-                None,
-                Some(VMError::FunctionCallError(FunctionCallError::MethodResolveError(
-                    MethodResolveError::MethodUTF8Error,
-                ))),
-            )
-        }
-    };
-    if let Err(e) = check_method(&module, method_name) {
-        return (None, Some(e));
+    run_vm(
+        code_hash,
+        code,
+        method_name,
+        ext,
+        context,
+        wasm_config,
+        fees_config,
+        promise_results,
+        VMKind::default(),
+    )
+}
+pub fn run_vm<'a>(
+    code_hash: Vec<u8>,
+    code: &[u8],
+    method_name: &[u8],
+    ext: &mut dyn External,
+    context: VMContext,
+    wasm_config: &'a VMConfig,
+    fees_config: &'a RuntimeFeesConfig,
+    promise_results: &'a [PromiseResult],
+    vm_kind: VMKind,
+) -> (Option<VMOutcome>, Option<VMError>) {
+    use crate::wasmer_runner::run_wasmer;
+    use crate::wasmtime_runner::run_wasmtime;
+    match vm_kind {
+        VMKind::Wasmer => run_wasmer(
+            code_hash,
+            code,
+            method_name,
+            ext,
+            context,
+            wasm_config,
+            fees_config,
+            promise_results,
+        ),
+        VMKind::Wasmtime => run_wasmtime(
+            code_hash,
+            code,
+            method_name,
+            ext,
+            context,
+            wasm_config,
+            fees_config,
+            promise_results,
+        ),
     }
+}
 
-    match module.instantiate(&import_object) {
-        Ok(instance) => match instance.call(&method_name, &[]) {
-            Ok(_) => (Some(logic.outcome()), None),
-            Err(err) => (Some(logic.outcome()), Some(err.into_vm_error())),
-        },
-        Err(err) => (Some(logic.outcome()), Some(err.into_vm_error())),
+pub fn with_vm_variants(runner: fn(VMKind) -> ()) {
+    for vm in vec![VMKind::Wasmer, VMKind::Wasmtime].iter() {
+        runner(vm.clone());
```

Comment:
TIL that Rust implements array-value iterator only under a [nightly feature-gate "array_value_iter"](https://github.com/rust-lang/rust/issues/65798). However, in this case you can do the following:

1. Mark `VMKind` as `Copy` type (`#[derive(Copy)`)
2. ```rust
    for vm in &[VMKind::Wasmer, VMKind::Wasmtime] {
        runner(*vm);
    }
    ```

This way you don't need to allocate Vec.

Reply:
Thanks!